### ISSUE TITLE
Transaction Duration Metrics don't match Measurement Metrics

### ISSUE THREAD

**Issue by kylannjohnson** [getsentry/sentry#68747]
### Environment

SaaS (https://sentry.io/)

### Steps to Reproduce

Consider two transactions, one un-scoped and the other scoped.  (Let's call them Transaction A and B).  Transaction A's duration is added as a measurement to Transaction B because we'd like to just track the duration of this operation instead of using a transaction on it.

The metrics don't seem to match and my question is whether they should or not.

In the dashboard below the left panels are showing various functions on `transaction.duration` of Transaction A.  The lowest panel is just a list of the duration.  The panels on the right are showing the same functions but applied to a measurement sent with Transaction B (the scoped one).

### Expected Result

I expect there's some interpolation going on in both cases, but I would expect the same results, given the same, small set of data.  Should a percentile calculation done on a measurement be the same as a duration of the same amount?   All durations are known and seem to both be in the data set.

If there is some variance, what is the accuracy of it?  e.g. +/- 1%

### Actual Result

<img width="1830" alt="Screenshot 2024-04-08 at 10 21 21 AM" src="https://github.com/getsentry/sentry/assets/1553804/85cc4816-c762-49e1-befc-aa1d8796eb4c">

### Product Area

Dashboards

┆Issue is synchronized with this [Jira Improvement](https://getsentry.atlassian.net/browse/FEEDBACK-2144) by [Unito](https://www.unito.io)

**Comment by k-fish** [comment]
Hey there, thanks for writing in, I think I understand what you're asking, but before we get into how the percentiles etc. work can you possibly show the events durations and the measurement in discover (or if you forgo using aggregates in dashboards and add event-id, the table should work) and list out each event? just to make sure the underlying numbers are the same?

eg. something like this (ignore that the measurement here is `num_of_spans` and replace it with the measurement you're using)
![Screenshot 2024-04-15 at 4 31 31 PM](https://github.com/getsentry/sentry/assets/6111995/087e1391-db80-4426-9918-1a72186f0d4e)

Also do you have any filters on these? I see `count()` of 8 for measurements and transactions, so if that's correct it should be showing the same number of data points.

**Comment by kylannjohnson** [comment]
Sorry, but the durations were obscured by the excel sheet and they matched.  They were listed in tables on the bottom with the only filter on the left being `title:TransactionA` and the right being `title:TransactionB`.  Then the columns were for `transaction.duration` and `measurement.foo` respectively

**Comment by kylannjohnson** [comment]
I was able to zero in on the same dataset 
![image](https://github.com/getsentry/sentry/assets/1553804/23564d65-6708-4a3b-ae60-46449d603332)

**Comment by kylannjohnson** [comment]
Sorry for the delay.  Let me make sure I can disclose that

**Comment by kylannjohnson** [comment]
> Can you possibly share the id for this dashboard (`sentry.io/dashboard/<id here>/`)? I can check on the backend and maybe figure out what's up.

ID: 85476

**Comment by gggritso** [comment]
@kylannjohnson thanks for providing the dashboard ID 👍🏻 We're taking a look, and will get back to you early next week–some of the folks working on this are away at the moment. Thank you for your patience!

**Comment by wmak** [comment]
Hmm this seems to be an issue with how we're storing this specific metrics data, i've reached out to the metrics team to help debug this. I'll let you know when I hear back from them.

**Comment by wmak** [comment]
Hey @kylannjohnson thanks for raising this you've helped us discover a pretty gnarly bug. This is our current understanding of what's happening
1. On April 8th, login_to_vitals measurements were sent to Sentry with a mix of units, some milliseconds, others seconds
2. Currently Sentry will always pick the first unit it sees and assume all the other values for a given measurement will use the same unit
3. This means that resulting aggregations are inconsistent with the sample list you have (eg. the 7.74s item in your original screenshot, is actually 7.74ms, you can confirm this if you open the event json for that in the Trace View)
<img width="222" alt="image" src="https://github.com/getsentry/sentry/assets/4205004/557ce521-d720-48b1-83c9-2043f52db068">

What this means for you hopefully is
1. Immediately - As long as you stick with a single unit, the data will be correct now
2. Short Term - We're going to fix the way we display samples to take multiple units into account
3. Long Term - We're going to introduce a way to handle multiple units more gracefully on storage so that when we can we'll convert the units to a consistent value

**Comment by kylannjohnson** [comment]
@wmak Is there any way to delete the current setting for this measurement (or any measurement)?  In this case, I think we would be fine starting from scratch because we're trying to move off of transaction.duration to a measurement.  

Is the selection of the first unit project-based?  Meaning, if we had a "staging project" that we tested on before moving to a production project, would that avoid this?

**Comment by wmak** [comment]
👋 My assumption here is we can't but I've reached out to @matejminar who has more information regarding the storage of measurements who can confirm for you.

**Comment by wmak** [comment]
Alternatively, if you just want to get the immediate solution you can rename the measurement too which should work immediately.

As well yes the unit determination is based on the currently selected projects

**Comment by matejminar** [comment]
@wmak is right, we can't delete [measurements](https://docs.sentry.io/platforms/javascript/tracing/instrumentation/performance-metrics/#custom-measurements). The best bet would be to use a different name if you'd like to start fresh.

slightly unrelated side note: we don't support deletion in custom metrics either; we just support "disabling" which blocks ingestion and hides the metric from UI, but unfortunately, that's not available with custom measurements
https://docs.sentry.io/product/explore/metrics/metrics-settings/#emitted-metrics

**Comment by kylannjohnson** [comment]
Ok, we'll explore versioning our metric/measurement names.  Looking forward to the fix, too.

**Comment by wmak** [comment]
>  I just wanted to be sure I understand. If we send a measurement to Project1 and then to Project2, is that considered two different measurements?

Not really, its more that per request we look at the selected projects and try our best to figure out what the user means when you request a measurement. 

If both projects are selected, the unit from either project may be used.

**Comment by kylannjohnson** [comment]
Hey all, any updates here?

**Comment by ale-cota** [comment]
Hi @kylannjohnson,
Sorry for the delayed reply here. Have you managed to implement the short term fix we suggested to rename the measurements and start fresh with a single unit? Are you still experiencing issues in your dashboard?

We don’t have an ETA yet for the longterm fix to handle and merge multiple units on the backend. This is currently on our backlog, but if you're still having issues, we'll try to find a solution. Thanks!

**Comment by kylannjohnson** [comment]
Hey @ale-cota ,

Thanks for the response.  Yes, we have decided to version our measurement names going forward.  It seems like this will address our issue in our local tests on a separate dashboard.

Perhaps there's already a way to do this, but some kind of Measurement Summary could be a way to indicate what raw values came in to Sentry or what unit they are being assigned.  I can see why the implementation is "sticky", but versioning a measurement name also isn't great.   I'm sure there's a great solution to this product level issue, but we'll keep an eye out as we grow our usage of measurements in the near future.

**Comment by ale-cota** [comment]
Thanks for confirming that versioning for the names at least addresses the issue for now. We'll update you once we manage to schedule the investigation work for merging/summarizing multiple units.

**Comment by bcoe** [comment]
@kylannjohnson in Sentry's new [Span Metrics](https://docs.sentry.io/platforms/javascript/guides/react/tracing/span-metrics/#custom-metrics) units are no longer a feature, with a small set of types being accept and formatting being applied at display time.

I believe given this new system, this issue is now stale.

Please let us know if you have any follow up questions.
